### ISSUE TITLE
Node url

### ISSUE THREAD

**Issue by PatrickMassot** [tiarno/plastex#33]
This is more a question than an issue but I haven't been able to find a better place to ask.

I'm lost in the plasTeX code and don't understand how rendering of ``\ref`` does its magic. It is able to get the url of the node referred to even if this node appear after the ``\ref`` in the TeX source (and the plasTeX DOM). How does it work?

The reason I'm asking is because I would like to be able, when rendering any node, to get the url of a another node. It seems hard, especially because of file splitting of the document, but ``\ref`` does it. 

I fear this question is obscure. Maybe a more specific question would help.
Say I would like to implement ``\cref`` from the [cleveref](http://www.ctan.org/tex-archive/macros/latex/contrib/cleveref) package. It allows to type "Using \cref{nice_theorem, useful_theorem}, we see..." and get "Using theorems 2 and 4, we see...".

But ideally I'd like to be able to get the url of any node, not necessarily having a label, as long as I can get a hand on its internal plasTeX object.

**Comment by kesmit13** [comment]
The location of `\ref`s don't matter because the whole document is read into memory into a DOM.  The code that determines what the filename and URL of a node in the DOM is in plasTeX/Renderers/__init__.py in the Renderable class.  The two properties to look at are `url` and `filename`.  This class is mixed into all DOM nodes when rendering happens, so you can use `url` and `filename` on any node to get that information.

**Comment by PatrickMassot** [comment]
Thank you very much. It's really weird, I was sure I tried this without luck but now it works. I guess I must have tried to invoke these properties after parsing but before starting the rendering process. 

The parsing phase is still a bit tricky since the deferred label resolution mechanism (in ``Context.ref``/``Context.label``) seems to be targeted at resolving a single label at a time. But it's easy to slightly modify the TeX class to allow nodes to register post-parsing callbacks. And this could be useful for other things I guess. I think I'll submit a PR after using it more.

I'd be grateful if you could use your power to intimidate plasTeX in order to also settle issues #28 and #29. A working ``newtheorem`` would be really really useful in order to use plasTeX for mathematics. Actually I find it interesting that you first fought so hard for tabular environment which would come sooo long after newtheorem in the wishlist of any LaTeX user that I know. It's interesting because it means there are more flavors of TeX users than I suspected.

**Comment by kesmit13** [comment]
This tool was actually written for our company's documentation system.  We don't use newtheorem, so that's why it hasn't really been tested.  We do however use a lot of tables.  Unfortunately, it may be a while before I can get to looking at this.
